packages/gatsby/src/utils/jobs-manager.js

```diff
@@ -0,0 +1,203 @@
+ * @param {Job} job
+ * @return Promise<T>
+ */
+const runLocalWorker = async (workerFn, job) => {
+  await fs.ensureDir(job.outputDir)
+
+  return new Promise((resolve, reject) => {
+    // execute worker nextTick
+    // TODO should we think about threading/queueing here?
+    process.nextTick(() => {
+      try {
+        Promise.resolve(
+          workerFn({
+            inputPaths: job.inputPaths,
+            outputDir: job.outputDir,
+            args: job.args,
+          })
+        ).then(resolve, reject)
+      } catch (err) {
+        reject(err)
+      }
+    })
+  })
+}
+
+/**
+ *
+ * @param {AugmentedJob} job
+ * @param {Job["plugin"]} plugin
+ * @param {pDefer.DeferredPromise} deferred
+ */
+const runJob = (job, plugin, deferred) => {
+  const worker = require(`${plugin.name}/worker.js`)
+  if (!worker[job.name]) {
+    deferred.reject(new Error(`No worker function found for ${job.name}`))
+    return
+  }
+
+  runLocalWorker(worker[job.name], job).then(deferred.resolve, deferred.reject)
+}
+
+const handleJobEnded = () => {
+  if (--activeJobs === 0) {
+    activityForJobs.end()
+    activityForJobs = null
+  }
+}
+
+/**
+ * Creates a job
+ *
+ * @param {Job} args
+ * @return {Promise<unknown>}
+ */
+exports.enqueueJob = async ({ name, inputPaths, outputDir, args, plugin }) => {
+  const rootDir = store.getState().program.directory
+
+  // TODO see if we can make this async, filehashing might be expensive to wait for
+  // currently this needs to be sync as we could miss jobs to have been scheduled and
+  // are still processing their hashes
+  const inputPathsWithContentDigest = inputPaths.map(path => {
+    return {
+      path: convertPathsToRelative(path, rootDir),
+      contentDigest: createFileHash(path),
+    }
+  })
+
+  const job = {
+    id: uuid(),
+    name,
+    inputPaths: inputPathsWithContentDigest,
+    outputDir: convertPathsToRelative(outputDir, rootDir),
+    args,
+    plugin,
+  }
+
+  job.contentDigest = createContentDigest({
+    name: job.name,
+    inputPaths: job.inputPaths.map(inputPath => inputPath.contentDigest),
+    outputDir: job.outputDir,
+    args: job.args,
+    plugin: job.plugin,
+  })
+
+  // When we already have a job that's executing, return the same promise.
+  if (jobsInProcess.has(job.contentDigest)) {
+    return jobsInProcess.get(job.contentDigest).deferred.promise
+  }
+
+  // Bump active jobs
+  activeJobs++
+  if (!activityForJobs) {
+    activityForJobs = reporter.phantomActivity(`Running jobs`)
+    activityForJobs.start()
+  }
+
+  // TODO: check cache folder for already stored jobs
+  const deferred = pDefer()
+  // node 8 doenst have finally so we call then & catch
+  deferred.promise = deferred.promise
+    .catch(err => {
+      handleJobEnded()
+
+      throw err
+    })
+    .then(res => {
+      handleJobEnded()
+
+      return res
+    })
+
+  jobsInProcess.set(job.contentDigest, {
+    id: job.id,
+    deferred,
+  })
+
+  runJob(job, plugin, deferred)
```

Comment:
we can't because we don't want to await here :) We want to queue jobs but don't care about the result here. It's up to the user/plugin if they would like to wait. If they don't wait we just wait for everything to complete before we stop our process.

Reply:
It'll be cleaner with the same end-result; a promise that resolves once the runJob call resolves. The only reason not to just return the promise here is because you want to cleanup, hence you await. From the caller pov it's the same deal.
I think...? :)

Reply:
Oh yeah, I see what you mean, I reread the code and you're right 👍 Thanks. It's easier to read now. I still need `pDefer` just to keep it consistent with the backward compatibility to old jobs reducer.

Reply:
Nice, this now makes me realize the pdefered is a little redundant since you already get a promise from the runJob call. So other comment